### DISCUSSION TITLE
How to properly call an overloaded function/method with parameter type in union?

### DISCUSSION THREAD

**Discussion by Azureblade3808** [microsoft/pylance-release#782] [Q&A]
Title of this post could be inaccurate, so feel free to change it.

Now, see code below:
```python
from typing import Union, overload

class A: pass
class B: pass

@overload
def f0(x: A) -> A:
    ...

@overload
def f0(x: B) -> B:
    ...

def f0(x: Union[A, B]) -> Union[A, B]:
    return x

@overload
def f1(x: A) -> A:
    ...

@overload
def f1(x: B) -> B:
    ...

def f1(x: Union[A, B]) -> Union[A, B]:
    # Error reported:
    #
    # No overloads for "f0(x)" match parameters
    #   Argument types: (A | B)
    #
    # Pylance(reportGeneralTypeIssues)
    return f0(x)
```

I'm pretty sure that in implementation of `f1`, calling `f0(x)` is valid, no matter whether `x` is an instance of either `A` or `B`. 

I'm also pretty sure that the type decision could become much more complicated if there were more than one simple parameter, therefore  I won't consider such behavior as a bug.

Adding some type-checks could mute the reported error, but it would harm the running efficiency.

So, are there some other good ways to have the work done without triggering a reported error?

**Comment by erictraut** [comment]
See [this issue](https://github.com/microsoft/pyright/issues/1298) for additional context.

You could define another overload for `f0`:
```python
@overload
def f0(x: Union[A, B]) -> Union[A, B]:
    ...
```

**Reply by Azureblade3808** [reply]
Oh, great. Thank you.
